### PR TITLE
[receiver/snowflakereceiver] add factory and scraper

### PR DESCRIPTION
**Pull request by shalper2**
**Description:** Implementing factory.go and scraper.go.

**Link to tracking Issue:** [14754](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14754)

**Testing:** Unit testing covering both components

**Documentation:** Documentation updated to cover a minor change to the metadata.yaml

### CODE REVIEW

File: receiver/snowflakereceiver/factory.go [hunk 1]
```diff
@@ -13,3 +13,64 @@
...
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/snowflakereceiver/internal/metadata"
+)
+
+const (
+	typeStr         = "snowflakemetrics"
```
Comment (dmitryax) [inline receiver/snowflakereceiver/factory.go:30]: This should be aligned with the module name which is `snowflake`

File: receiver/snowflakereceiver/factory.go [hunk 2]
```diff
@@ -13,3 +13,64 @@
...
+		Database: defaultDB,
+		Role:     defaultRole,
+	}
+}
+
+func NewFactory() component.Factory {
+	return receiver.NewFactory(
+		typeStr,
+		createDefaultConfig,
+		receiver.WithMetrics(createMetricsReceiver, component.StabilityLevelAlpha),
```
Comment (dmitryax) [inline receiver/snowflakereceiver/factory.go:53]: Let's keep the stability level as Development as it's documented in the README. We can bump it as a separate follow-up PR

File: receiver/snowflakereceiver/scraper.go [hunk 3]
```diff
@@ -13,3 +13,216 @@
...
+	return &snowflakeMetricsScraper{
+		settings: settings.TelemetrySettings,
+		conf:     conf,
+		mb:       metadata.NewMetricsBuilder(conf.Metrics, settings),
+	}
+}
+
+// for use with receiver.scraperhelper
+func (s *snowflakeMetricsScraper) start(_ context.Context, _ component.Host) (err error) {
+	s.client, err = newDefaultClient(s.settings, *s.conf)
```
Comment (dmitryax) [inline receiver/snowflakereceiver/scraper.go:47]: I think we also need a shutdown function closing the database connection

File: receiver/snowflakereceiver/scraper.go [hunk 4]
```diff
@@ -13,3 +13,216 @@
...
+		s.mb.RecordSnowflakeRowsUnloadedAvgDataPoint(t, row.avgRowsUnloaded, row.attributes.schemaName.String, row.attributes.executionStatus.String, row.attributes.errorMessage.String, row.attributes.queryType.String, row.attributes.warehouseName.String, row.attributes.databaseName.String, row.attributes.warehouseSize.String)
+		s.mb.RecordSnowflakeRowsUpdatedAvgDataPoint(t, row.avgRowsUpdated, row.attributes.schemaName.String, row.attributes.executionStatus.String, row.attributes.errorMessage.String, row.attributes.queryType.String, row.attributes.warehouseName.String, row.attributes.databaseName.String, row.attributes.warehouseSize.String)
+		s.mb.RecordSnowflakeTotalElapsedTimeAvgDataPoint(t, row.avgTotalElapsedTime, row.attributes.schemaName.String, row.attributes.executionStatus.String, row.attributes.errorMessage.String, row.attributes.queryType.String, row.attributes.warehouseName.String, row.attributes.databaseName.String, row.attributes.warehouseSize.String)
+
+		s.mb.EmitForResource(
+			metadata.WithSnowflakeAccountName(s.conf.Account))
```
Comment (dmitryax) [inline receiver/snowflakereceiver/scraper.go:176]: It shouldn't be called if all the resources are the same. Just call `s.mb.Emit(metadata.WithSnowflakeAccountName(s.conf.Account))` once at the end of `scrape`

Comment (shalper2) [reply receiver/snowflakereceiver/scraper.go:176]: I think I misunderstood the definition of a resource, but this makes sense on re-read of definitions.

File: receiver/snowflakereceiver/factory.go (hunk 1 above)
Comment (shalper2) [reply receiver/snowflakereceiver/factory.go:30]: heard

File: receiver/snowflakereceiver/factory.go (hunk 2 above)
Comment (shalper2) [reply receiver/snowflakereceiver/factory.go:53]: Sorry I did not see development as an enum on my editor! That's what I get for being lazy.

File: receiver/snowflakereceiver/scraper.go [hunk 5]
```diff
@@ -13,3 +13,197 @@
...
+	s.client, err = newDefaultClient(s.settings, *s.conf)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *snowflakeMetricsScraper) shutdown(_ context.Context) (err error) {
+	err = s.client.client.Close()
+	return err
```
Comment (dmitryax) [inline receiver/snowflakereceiver/scraper.go:56]: It's not always guaranteed that shutdown is called after the start succeeds. So we need an extra check here
```suggestion
func (s *snowflakeMetricsScraper) shutdown(_ context.Context) (error) {
	if s.client == nil {
		return nil
	}
	return s.client.client.Close()
```

**Review by dmitryax — COMMENTED** [review]
Just one nit. Otherwise LGTM

File: receiver/snowflakereceiver/scraper.go (hunk 5 above)
Comment (shalper2) [reply receiver/snowflakereceiver/scraper.go]: got it, makes sense to me!

**Review by dmitryax — APPROVED** [review]
LGTM

**Pull request merged**